src/algebra/Module/basic.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2019 Robert A. Spencer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert A. Spencer
+-/
+import algebra.module
+import algebra.punit_instances
+import category_theory.types
+import linear_algebra.basic
+open category_theory
+
+universe u
+
+variables (R : Type u) [ring R]
+
+/-- The category of R-modules and their morphisms. -/
+structure Module :=
+  (carrier : Type u)
+  [is_add_comm_group : add_comm_group carrier]
+  [is_module : module R carrier]
+
+attribute [instance] Module.is_add_comm_group Module.is_module
+
+namespace Module
+
+instance : has_coe_to_sort (Module R) :=
+  { S := Type u, coe := Module.carrier }
+
+def of (X : Type u) [add_comm_group X] [module R X] : Module R := ⟨R, X⟩
```

Comment:
I wonder whether we should try to get away with making `R` implicit here, since it is determined by the return type. Anyone have thoughts?

Reply:
My thoughts are that making R explicit is good policy.  Any R module is a module for any subring of R, the integers, the `punit` etc etc.  I'd like to be clear in all cases.

Reply:
Let's leave it explicit for now. I'd rather have a few more `_`s than a bunch of `@`s all over the place.

Reply:
I've found an interesting use case where making R implicit is required mathimatically (and think it motivates it being required everywhere)

Suppose we have a definition of triviality with an implicit ring
``` lean
def are_isomorphic {R} [ring R] (M₁ M₂ : Module R) := nonempty (M₁ ≅ M₂)
def is_trivial (M : Module R) := are_isomorphic M 0
```
Then the following does not work
```lean
variables (R : Type) [ring R] (M : Module R) (S : submodule R M)
#check is_trivial S
```
as it is unable to synthesize the ring variable.  You have to put `@is_trivial R _ S` to get it to work.

I think being explicit is desired, because S _is_ trivial as, say, a module over the 0 ring, and all R modules are modules over the 0 ring.

Reply:
This `is_trivial` example is more related to the behavior of coercions. The coercion `has_coe (submodule R M) (Module R)` will only apply when Lean already knows that it needs to coerce a `submodule R M` to a `Module R`. Here we just know we need to coerce `submodule R M` to `Module R'` for some `R'`, so the coercion doesn't apply.

If you had an ordinary function `of_submodule : submodule R M -> Module R` and applied it explicitly, then you wouldn't need to specify the ring variable in `is_trivial (of_submodule S)`.

Alternatively, using the coercion, you can also write `is_trivial (S : Module R)`.

Reply:
That all makes sense.

The discussion around `is_trivial` and such forth is for a later time, I suppose.  Leaving explicit is probably best policy.